### PR TITLE
20171127 login screen

### PR DESCRIPTION
**Pull request by at-thinhnguyen2**

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 1]
```diff
@@ -2,22 +2,33 @@
...
+//  Created by Thinh Nguyen X. on 11/29/17.
+
 //  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
 //
 
 import UIKit
 
 final class HomeViewController: UIViewController {
+
+    @IBOutlet fileprivate weak var logoutButton: UIButton!
```
Comment (at-daiho) [inline MyApp/View/Controller/Home/HomeViewController.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 2]
```diff
@@ -2,22 +2,29 @@
...
+//  Created by Thinh Nguyen X. on 11/29/17.
 //  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
 //
 
 import UIKit
 
 final class HomeViewController: UIViewController {
+
+    @IBOutlet weak var logoutButton: UIButton!
+
```
Comment (at-daiho) [inline MyApp/View/Controller/Home/HomeViewController.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/ViewModel/LoginViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            case .success: return "Success"
+            case .failure(field: _, let msg):
+                return "Failure: " + msg
+            }
+        }
+    }
+
+    var username = ""
+    var password = ""
+
```
Comment (at-daiho) [inline MyApp/ViewModel/LoginViewModel.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/ViewModel/LoginViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            case .success: return "Success"
+            case .failure(field: _, let msg):
+                return "Failure: " + msg
+            }
+        }
+    }
+
+    var username = ""
+    var password = ""
+    var userlog: [String: Any] = [ : ]
```
Comment (at-daiho) [inline MyApp/ViewModel/LoginViewModel.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/ViewModel/LoginViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                return "Failure: " + msg
+            }
+        }
+    }
+    var username = ""
+    var password = ""
+    var userlog: [String: Any] = [ : ]
+    var userlogCheck: [String: Any] =  [ : ]
+    func validate() -> Validation {
+        guard username.len >= 6 else {
```
Comment (at-daiho) [inline MyApp/ViewModel/LoginViewModel.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/ViewModel/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                return "Failure: " + msg
+            }
+        }
+    }
+    var username = ""
+    var password = ""
+    var names: [String] = []
+    var passwords: [String] = []
+    func validate() -> Validation {
+        guard username.len >= 6 else {
```
Comment (at-daiho) [inline MyApp/ViewModel/LoginViewModel.swift]: WARNING - Lines should not have trailing whitespace.

**Comment by at-daiho** [comment]
Total linter violations in pull request: 6
Serious: 0
Warnings: 6

**Comment by at-daiho** [comment]
Total linter violations in pull request: 3
Serious: 0
Warnings: 3

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 7]
```diff
@@ -2,22 +2,32 @@
...
+//  Created by Thinh Nguyen X. on 11/29/17.
+
 //  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
 //
 
 import UIKit
 
 final class HomeViewController: UIViewController {
+    @IBOutlet fileprivate weak var logoutButton: UIButton!
+
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 8]
```diff
@@ -2,22 +2,32 @@
...
-//  Created by Thinh Nguyen X. on 11/24/17.
+//  Created by Thinh Nguyen X. on 11/29/17.
+
 //  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
 //
 
 import UIKit
 
 final class HomeViewController: UIViewController {
+    @IBOutlet fileprivate weak var logoutButton: UIButton!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: add linespace, mark

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 9]
```diff
@@ -2,22 +2,32 @@
...
+    // MARK: - Life cycle
     override func viewDidLoad() {
         super.viewDidLoad()
-        view.backgroundColor = App.Color.red214
-        checkTheFistRun()
+        setupActions()
+    }
+}
+
+extension HomeViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: di chuyển lên mấy hàm ni lên trên kia luôn

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    @IBOutlet fileprivate weak var loginButton: UIButton!
+
+    // MARK: - Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupUI()
+        setupActions()
+    }
+
+    func setupUI() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        passwordTextField.isSecureTextEntry = true
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+        view.addGestureRecognizer(tap)
+        passwordTextField.delegate = self
+        usernameTextField.delegate = self
+        usernameTextField.returnKeyType = .next
+        passwordTextField.returnKeyType = .done
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: remove nếu ko dùng đến

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 12]
```diff
@@ -2,22 +2,32 @@
...
 //
-//  Created by Thinh Nguyen X. on 11/24/17.
+//  Created by Thinh Nguyen X. on 11/29/17.
+
 //  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
 //
 
 import UIKit
 
 final class HomeViewController: UIViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: `final class HomeViewController: ViewController `

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+//  MyApp
+//
+//  Created by Thinh Nguyen X. on 11/27/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import UIKit
+import MVVM
+
+final class LoginViewController: UIViewController, MVVM.View {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]:
```swift
final class LoginViewController: ViewController, MVVM.View
```

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        if passwordTextField.isFirstResponder { return }
+        usernameTextField.becomeFirstResponder()
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        view.endEditing(true)
+    }
+
+    func textField(for field: LoginViewModel.Field) -> UITextField {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        switch field {
+        case .username: return usernameTextField
+        case .password: return passwordTextField
+        }
+    }
+
+    @objc func dismissKeyboard() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        view.endEditing(true)
+    }
+}
+
+// MARK: - Private
+extension LoginViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: đưa những hàm này lên trên. Extension chỉ khi có kế thừa từ đối tượng nào đó

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+    fileprivate func setupActions() {
+        loginButton.addTarget(self, action: #selector(login), for: .touchUpInside)
+    }
+
+    @objc func login() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+                   this.alert(error: error)
+                }
+                this.viewDidUpdated()
+            }
+        case .failure(let field, let msg):
+            showAlertError(msg: msg, field: field)
+        }
+    }
+
+    func showHome() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        let homeVC = HomeViewController()
+        if let navi = navigationController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: navigationController?.pushViewController

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+            navi.pushViewController(homeVC, animated: true)
+        }
+    }
+
+    func showAlertError(msg: String, field: LoginViewModel.Field) {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        alert(title: App.String.error, msg: msg, buttons: [App.String.ok], handler: { [weak self] _ in
+            guard let this = self else { return }
+            let fld = this.textField(for: field)
+            fld.becomeFirstResponder()
+        })
+        viewDidUpdated()
+    }
+
+    func updateView() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: private func

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        guard isViewLoaded else { return }
+        usernameTextField.text = viewModel.username
+        passwordTextField.text = viewModel.password
+    }
+}
+
+// MARK - TextField Delegate
+extension LoginViewController: UITextFieldDelegate {
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        if textField == self.usernameTextField {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: remove self

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+            passwordTextField.becomeFirstResponder()
+        } else if textField == self.passwordTextField {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: remove self

File: MyApp/ViewModel/LoginViewModel.swift [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        var description: String {
+            switch self {
+            case .success: return "Success"
+            case .failure(field: _, let msg):
+                return "Failure: " + msg
+            }
+        }
+    }
+    var username = ""
```
Comment (at-linhvo) [inline MyApp/ViewModel/LoginViewModel.swift]: add linespace

File: MyApp/ViewModel/LoginViewModel.swift [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    var password = ""
+    func validate() -> Validation {
```
Comment (at-linhvo) [inline MyApp/ViewModel/LoginViewModel.swift]: add linespace

File: MyApp/ViewModel/LoginViewModel.swift [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+        return .success
+    }
+
+    enum LoginResult {
+        case success
+        case failure(Error)
+    }
+
+    func login(_ completion: @escaping (LoginResult) -> Void) {
+        let validation = self.validate()
```
Comment (at-linhvo) [inline MyApp/ViewModel/LoginViewModel.swift]: remove self

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

**Comment by at-linhvo** [comment]
LGTM!

**Pull request merged**